### PR TITLE
Add tracing sdk to function runtimes

### PR DESCRIPTION
**Pull request by pPrecel**
**Description**

Changes proposed in this pull request:

- Extend all function runtimes by adding the tracer object

**Related issue(s)**

- https://github.com/kyma-project/kyma/issues/11327

### CODE REVIEW

**Comment by pPrecel** [comment]
I've tested all functionality using the [http-https-echo](https://hub.docker.com/r/mendhak/http-https-echo) to receive and display information about requests outgoing from the serverless functions and the KymaCLI to easily deploy each runtime.

1. To configure `http-https-echo` service apply the following yaml:

echo.yaml
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: my-echo
  name: my-echo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-echo
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: my-echo
    spec:
      containers:
      - image: mendhak/http-https-echo:23
        name: http-https-echo
        resources: {}
status: {}
---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: my-echo
  name: my-echo
spec:
  ports:
  - name: 8080-8080
    port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    app: my-echo
  type: ClusterIP
status:
  loadBalancer: {}
```

2. For the nodejs12/14 runtimes:

config.yaml
```yaml
name: function-exciting-rafal8
namespace: default
runtime: nodejs14
source:
    sourceType: inline
apiRules:
    - rules:
        - methods:
            - GET
            - POST
            - PUT
            - PATCH
            - DELETE
            - HEAD
```

handler.js
```javascript
const api = require('@opentelemetry/api');

module.exports = {
    main: function (event, context) {
        console.log('starting...');
        req = event.extensions.request;

        if (req.headers['x-b3-sampled'] == 0) {
            console.log("sampled value is equal 0... skipping :(")
            return
        }

        console.log("headers:", req.headers);
        
        span = event.tracer.startSpan('my-first-own-span', {});
        console.log("span:", span);
        span.addEvent('bar');

        for (let i = 0; i < 10000000; i++){};
        
        span2 = event.tracer.startSpan('my-sec-own-span', {});
        console.log("span2:", span2);

        for (let i = 0; i < 10000000; i++){};

        span2.addEvent('good job');

        api.context.with(api.setSpan(api.context.active(), span), () => {
            span3 = event.tracer.startSpan('my-sec-of-the-first-own-span', {
            });

            console.log("span3:", span2);

            var XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;
            const Http = new XMLHttpRequest();
            const url='http://my-echo.default.svc.cluster.local:8080';
            Http.open("GET", url);
            Http.send();
            Http.onreadystatechange = (e) => {
                console.log(Http.responseText)
            }

            for (let i = 0; i < 10000000; i++){}

            span3.addEvent('good job');
            span3.end();
        })

        for (let i = 0; i < 10000000; i++){};

        span2.end();
        
        span.addEvent('almost end');
        span.end();

        console.log("tracing initialized");
        
        return 'Hello Serverless'
    }
}
```

package.json
```json
{
  "name": "function-exciting-rafal8",
  "version": "0.0.1",
  "dependencies": {
    "@opentelemetry/api": "^0.18.0",
    "xmlhttprequest": "^1.8.0"
  }
}
```

3. Dor the python39 runtime:

config.yaml
```yaml
name: function-epic-maciej1
namespace: default
runtime: python39
source:
    sourceType: inline
apiRules:
    - rules:
        - methods:
            - GET
            - POST
            - PUT
            - PATCH
            - DELETE
            - HEAD
```

handler.py
```python
import requests


def main(event, context):
    ce = event.ceHeaders["extensions"]
    
    if ce["request"].headers.get('X-B3-Sampled') == "0":
        print("sampled 0 :( skipping")
    else:
        with event.tracer.start_as_current_span("fir-span"):
            event.tracer.start_span("fir-of-fir-span")
            with event.tracer.start_as_current_span("sec-of-fir-span"):
                print("Hello OpenTelemetry with Zipkin from the sec-of-fir-span!")
                r = requests.get('http://my-echo.default.svc.cluster.local:8080')
                print(r.text)
        with event.tracer.start_as_current_span("sec_span"):
            print("hello from sec span")

    return "hello world" 
```

**Review by Cortey — APPROVED** [review]

**Review by dbadura — CHANGES_REQUESTED** [review]
I revieweed the Python part

File: components/function-runtimes/python39/kubeless/tracing.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+from opentelemetry.trace.propagation import _SPAN_KEY
+from typing import Iterator
+
+podName = os.getenv('HOSTNAME')
+serviceNamespace = os.getenv('SERVICE_NAMESPACE')
+jaegerEndpoint = os.getenv('JAEGER_SERVICE_ENDPOINT')
+
+def is_jaeger_available() -> bool:
+    try:
+        res = requests.get(jaegerEndpoint)
```
[the remark this answers is not in the document]
Comment (pPrecel) [reply components/function-runtimes/python39/kubeless/tracing.py:24]: good point

File: components/function-runtimes/python39/kubeless/tracing.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    
+        # 405 is the right status code for the GET method if jaeger service exists
```
[the remark this answers is not in the document]
Comment (pPrecel) [reply components/function-runtimes/python39/kubeless/tracing.py:26]: `405` status means that [Method Not Allowed](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405). There is no info in the documentation that this endpoint would return this status. But this results from the HTTP servers architecture. If the method would be not allowed, then the server will return the `405` status code. Here is info about the POST method which is the only allowed:
https://github.com/jaegertracing/jaeger/blob/7872d1b07439c3f2d316065b1fd53e885b26a66f/cmd/collector/app/handler/http_handler.go#L60

[the remark this answers is not in the document]
Comment (dbadura) [reply components/function-runtimes/python39/kubeless/tracing.py:26]: I would add description of this to the code to minimize confusion

[the remark this answers is not in the document]
Comment (dbadura) [reply components/function-runtimes/python39/kubeless/tracing.py:26]: I think we should we should use `health` endpoint: 
https://github.com/jaegertracing/jaeger/blob/7872d1b07439c3f2d316065b1fd53e885b26a66f/pkg/healthcheck/handler.go

**Review by dbadura — DISMISSED** [review]
I reviewed python part

**Review by dbadura — APPROVED** [review]

**Review by Cortey — APPROVED** [review]

**Pull request merged**